thunder/executors/transformer_engine_v2ex_impl.py

```diff
@@ -0,0 +1,390 @@
+from thunder.executors.transformer_engineex import _linear_checker
+import thunder.torch as ltorch
+from thunder import Transform
+from thunder.core import prims
+from thunder.core.proxies import Proxy, Variable, unvariableify, variableify
+from thunder.core.trace import from_trace, TraceProvenance, TraceTag
+from thunder.core.transforms import (
+    _update_forward_with_new_saved_for_backward,
+    _update_backward_with_new_saved_for_backward,
+)
+from thunder.core.transform_common import cse_single_bsym
+from thunder.executors.passes import del_last_used
+
+if TYPE_CHECKING:
+    from thunder.core.trace import VariableInterface
+    from thunder.core.symbol import BoundSymbolRHS, BoundSymbol
+    from thunder.core.proxies import TensorProxy
+
+import transformer_engine.pytorch as te
+from transformer_engine.pytorch.tensor.float8_tensor import Float8Quantizer
+from transformer_engine.pytorch.ops import BasicLinear
+from transformer_engine.pytorch.fp8 import (
+    _amax_and_scale_update,
+    get_default_fp8_recipe,
+    get_fp8_max,
+    Recipe,
+    RecipeState,
+    FP8GlobalStateManager,
+)
+
+
+functional_te_ex = StatefulExecutor("functional_te")
+register_executor(functional_te_ex)
+
+
+class RecipeType(Enum):
+    DELAYED = auto()
+    MXFP8 = auto()
+    FP8_CURRENT_SCALING = auto()
+    FP8_BLOCK_SCALING = auto()
+
+
+def get_fp8_recipe_type(recipe: Recipe) -> RecipeType | None:
+    if recipe.delayed():
+        return RecipeType.DELAYED
+    if recipe.mxfp8():
+        return RecipeType.MXFP8
+    if recipe.float8_current_scaling():
+        return RecipeType.FP8_CURRENT_SCALING
+    if recipe.float8_block_scaling():
+        return RecipeType.FP8_BLOCK_SCALING
+    return None
+
+
+def _te_fp8_recipe_meta() -> AnyProxy:
+    return AnyProxy(None, prefix="r")
+
+
+class TERecipe:
+    def __init__(self):
+        self.fp8_recipe = None
+
+    def __call__(self) -> list[Recipe]:
+        # If the Thunder trace is not ran inside an fp8_autocast region then
+        # use the default recipe.
+        if not FP8GlobalStateManager.FP8_RECIPE:
+            if not self.fp8_recipe:
+                self.fp8_recipe = get_default_fp8_recipe()
+
+            return self.fp8_recipe
+
+        return FP8GlobalStateManager.get_fp8_recipe()
+
+
+_get_te_fp8_recipe = functional_te_ex.register_stateful_operator(
+    "get_te_fp8_recipe", meta=_te_fp8_recipe_meta, state_class=TERecipe
+)
+
+
+def _get_te_fp8_quantizers_meta(recipe_state: RecipeState, num_quantizers: int):
+    # num_quantizers can be taken from recipe state, but not at trace time.
+    # Therefore it's needed here to inform the meta.
+    return (*(AnyProxy(None, prefix="q") for _ in range(num_quantizers)),)
+
+
+class TEQuantizerState:
+    def __init__(self):
+        self.quantizers = None
+        self.parent_recipe_state: None | RecipeState = None
+
+    def __call__(self, recipe_state: RecipeState, num_quantizers: int) -> list[Float8Quantizer]:
```

Comment:
Depending on the recipe it could be other quantizer as well.

https://github.com/NVIDIA/TransformerEngine/blob/0e45e138c08af8f3b38e46eea58e2e9dbe628d42/transformer_engine/pytorch/tensor/mxfp8_tensor.py#L24

https://github.com/NVIDIA/TransformerEngine/blob/0e45e138c08af8f3b38e46eea58e2e9dbe628d42/transformer_engine/pytorch/tensor/float8_blockwise_tensor.py#L21

I think annotating with their base class `Quantizer` would be appropriate.